server/util/rexec/rexec.go

```diff
@@ -0,0 +1,452 @@
+// package rexec provides a command interface similar to the "os/exec" package
+// but runs commands using a remote execution client.
+//
+// Basic usage: you can just call Result() to run a command end-to-end,
+// including uploading the Action, Command, and input root dir to cache:
+//
+//	cmd := rexec.Command(env, "echo", "hello")
+//	res, err := cmd.Result(ctx)
+//	if err != nil {
+//		return err
+//	}
+//	fmt.Println(string(res.Stdout)) // prints "hello"
+//
+// You can also manually walk the command through its lifecycle:
+//
+//	// TODO: handle errors!
+//	cmd := rexec.Command(env, "echo" "hello")
+//	err := cmd.Upload(ctx)
+//	err = cmd.Start(ctx)
+//	err = cmd.Wait(ctx)
+//	res, err := cmd.Result(ctx)
+//
+// Command properties are configured by setting attributes on *Cmd:
+//
+//	cmd.Dir = "/path/to/input_root_dir"
+//	cmd.Env = []string{"FOO=BAR"}
+//	cmd.Platform = []string{"recycle-runner=true"}
+package rexec
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/retry"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/genproto/googleapis/longrunning"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	gstatus "google.golang.org/grpc/status"
+)
+
+const (
+	defaultDigestFunction = repb.DigestFunction_SHA256
+)
+
+type state int
+
+const (
+	initial state = iota
+	uploading
+	starting
+	waiting
+)
```

Comment:
I'm a little unclear about the expected use of the API -- is it safe to call Start/Upload/Wait multiple times? What about ordering?

I think a possibly clearer way to handle this would be to return errors from each of these methods if the state is not what you'd expect from normal operation.

Reply:
N/A after refactoring